rails generator: take other rodauth plugin translations into account

HoneyryderChuck opened the pull request:
In order to take rodauth plugin registered translations into account, the following is done:

1. store rodauth-only translation files in a global store
2. use it when generating the destinationn files

@janko I'm having an issue that requires your input. So, using generators might not cut it, as these do not load the application, so there's nothing in the registry to be picked up. This could alternatively be accomplished using a rails/rake task that loads the environment. Would that be an acceptable workaround, given that the cli interface would have to change?

Other suggestions are welcome.

HoneyryderChuck committed:
create global registry to set all rodauth i18n files

HoneyryderChuck committed:
copy all rodauth plugin translations to same destination rails locale
file

janko commented:
Thank you for working on this 🙏

I also thought about the issue of Rodauth app not being loaded. I think the Rails app is initialized in generators, but the Rodauth app is not autoloaded, as nothing is referencing it.

We could rely on rodauth-rails being used, and call `Rodauth::Rails.app` to ensure the Rodauth app is loaded. What do you think?

HoneyryderChuck commented:
Thx for the tip, totally forgot about setting up `rodauth-rails`, makes total sense.

>  I think the Rails app is initialized in generators,

I don't think it is :/ Calling `Rodauth::Rails.app` raise a `NameError: uninitialized constant Rodauth::Rails`. I really think this is a feature of generators, unless you can point me to an example of one stating otherwise.

janko commented:
That's interesting, I don't know how rodauth-rails' generators are then able to reference the Rodauth app 🤔 I will take a closer look.

janko commented:
Referencing `Rodauth::Rails.app` in a rodauth-i18n generators works for me on the official Rails demo app.

HoneyryderChuck commented:
just pushed the latest changes. Can you check if this is what you're doing to reference the app?

EDIT: I think I understood. It all works in `rodauth-rails` because there's a test rails app, which gets loaded in test-helper.

HoneyryderChuck commented:
@janko I think that's it!

It's a bit of a hassle that we need a full-fledged rails test app just to do this generator bit, but oh well 🤷 

Let me know what you think of the implementation.

janko commented:
The implementation looks mostly good, I have few comments:

1. I don't think we need to filter locale files based on filename, we can probably just load them all into one big hash, and import only the selected locales.
2. It might be simpler to use a constant instead of a class method for I18n files (e.g. `I18N_FILES`).
3. The Rodauth app will not get loaded automatically inside the generator, only once it's referenced.
4. Do we want to handle merging of new translations into existing ones?

Regarding the tests, I agree it would be useful to have the generator tested, but I'd also be fine if we didn't have them. In the case we will have them, we should probably strip down any unnecessary functionality/files. I could help with that if you want.

BTW, I pushed some recent changes to the generator, changing its location. I explained the reasoning in the commit description.

HoneyryderChuck commented:
rebased and implemented some of the changes.

1. this could grow unwieldy given the right amount of translations. Although I agree that perhaps there are better ways to do it, perhaps this could be done later?
2. I didn't do it explicitly, as I consider constants to be better suited for static and frozen values, and a bit of a "smell" to manipulate them in hooks. Let me know if I'm being too prescriptive.
3. Eager-loading it now.
4. Done (preloading them before the recursive merges).

Agree about stripping down functionality. Let me know what you think I could remove further.

janko commented:
1. I see your point, makes sense, let's keep the current implementation.
2. That's a fair point. My main motivation was easier initialization and access, but I agree that modifying constants isn't ideal. Though `Rodauth::FEATURES` technically doesn't follow this 🙂 
3. 👍🏻 
4. Looks good 👍🏻 Note that reading existing translations might not work when invoking the generator outside of the project root. It seems that Thor has `destination_root` you can use to generate an absolute path. Technically `Rails.root` might work too, but maybe `destination_root` is different in some cases?

Regarding removing cleaning up the test Rails app, we should be able to do the following:

* remove `Admin::RodauthController`
* clear contents of `RodauthController`
* remove `TestController` and routes
* define rodauth configuration directly in `RodauthApp` (since it's so simple)
* remove all view templates (we can use Rodauth's built-in ones)
* remove Active Record (since we don't need to test integration with it)
* remove Sequel initializer (we're already connecting in test helper)
* remove Active Record migration (we're already creating accounts table in test helper)

janko commented:
Also, when you're dumping data into YAML, you'll want to remove the 80-character column wrapping and the `---` header:

```rb
content = YAML.dump({ locale => data }, line_width: 1000)
content.split("\n", 2).last # remove YAML header
```

janko commented:
Regarding selecting translation files based on their filename, I realized the current implementation relies that the files will be named `{locale}.yml`. However, what if some gem provides multiple large Rodauth features, and they want to split translations into `feature1.{locale}.yml` and `feature2.{locale}.yml`? I would rather not make this assumption if we don't have to. I'm OK with assuming the locale will be part of the filename, so it should be fine to assume the filename *ends* in `{locale}.yml`.

janko commented:
Nevermind, we're already making this assumption when adding translation files I18n's load path, so we can keep it like it is 👍

HoneyryderChuck committed:
fix jruby

HoneyryderChuck commented:
hey @janko , I think I applied all the required and suggested changes.

> Though Rodauth::FEATURES technically doesn't follow this 

I think I modelled it around the `@features` class ivar under `Rodauth::Auth`, which may get frozen. Depending of what we'd want for this variable, things can change. I'll leave it up to you to decide :) As a 1st implementation, one's allowed to compromise here.

Regarding the test failures,  I addressed jruby, but rails 7 is apparently not compatible with ruby 3.1 yet.

janko commented:
Thanks for the changes. I will merge this and make several more modifications to it. After that I plan to modify the generator to only import translations for currently loaded features, which will become possible with this feature.

Yeah, I also saw that Rails 7 isn't compatible with Ruby 3.1 yet.
